### ISSUE TITLE
Autocomplete / AJAX on select2 datagrid filters

### ISSUE THREAD

**Issue by webdevilopers** [sonata-project/SonataAdminBundle#2360]
Unfortunately the docs are still missing for filters:
http://sonata-project.org/bundles/admin/master/doc/reference/action_list.html#filters

Is there a way to use `select2` with `AJAX` `autocompletion` on `datagridFilters` like the `sonata_type_model_autocomplete` type on `formFields`?

Sometimes I have a handful of filters and each select includes hundreds of rows.
Though cached this can become an overload.

Possibly related issues:
* https://github.com/sonata-project/SonataAdminBundle/pull/1795

**Comment by pulzarraider** [comment]
@webdevilopers @krewetka I have refactored the way how the data was transformed. The issues with pre-defined filters and missing allow clear should work now.
Thank you for the feedback. Please report any problems you will find.

**Comment by webdevilopers** [comment]
Awesome ! Gonna check it out next week. Any specific branch?

**Comment by pulzarraider** [comment]
It's still the same branch (sonata-project/SonataAdminBundle#2454). I have already squashed the commits.

**Comment by webdevilopers** [comment]
Great work @pulzarraider , the filters work as expected now!

I tried some pre-setting variations by leaving out the `type` or even hiding it via `'operator_type' => 'hidden'`.
In both cases the `TYPE_IS_EQUAL` seems to be used by default. is this correct and should it be noted in the docs later?

**Comment by pulzarraider** [comment]
Thank you @webdevilopers 

> I tried some pre-setting variations by leaving out the type or even hiding it via 'operator_type' => 'hidden'.
In both cases the TYPE_IS_EQUAL seems to be used by default. is this correct and should it be noted in the docs later? 
`TYPE_IS_EQUAL` is used by default. It's the expected behaviour, because if you filter something, you have to select only values that contains the specified value.

> The problem is that the type select2 is not populated by default and stays blank though it used the default filtering.
Type is blank by dafult in all filters.

**Comment by loconox** [comment]
In echo to @krewetka post, I'm encountering an issue with filters on nested resources.

I think it's linked with the fact that dot in filters form fields name are not allowed and are replaced with '__'. Wherees, in the HelperController the filter field is search by the form name and not by the filter name.
```
 private function retrieveFormFieldDescription(AdminInterface $admin, $field)
    {
        $admin->getFormFieldDescriptions();

        $fieldDescription = $admin->getFormFieldDescription($field);
...
```
Ex:
I create a new filter as follow:
```
 protected function configureDatagridFilters(DatagridMapper $datagridMapper)
    {
        $datagridMapper
            ->add('foo.bar', 'doctrine_orm_model_autocomplete', [], null, ['property' => 'name'])
            ;
    }

```
The filter form field will have the name 'foo__bar' and the $field value in the fonction 'getFormFieldDescription($field)' will be 'foo__bar' and not 'foo.bar', which is the current key value of the filter which interests us, in the array 'filterFieldDescriptions' in the admin object.

Hack solution :
Override the method getFilterFieldDescription in the CustomAdmin class :
```
    public function getFilterFieldDescription($name)
    {
        $nestedFieldName = str_replace('__', '.', $name);

        return $this->hasFilterFieldDescription($name) ? $this->filterFieldDescriptions[$name] : $this->hasFilterFieldDescription($nestedFieldName) ? $this->filterFieldDescriptions[$nestedFieldName] : null;
    }
```

**Comment by pulzarraider** [comment]
@loconox the bug with nested filter was fixed with commit ae2d48ef647a3bb4849cf5818e6297ff41f44a19

**Comment by pulzarraider** [comment]
I have merged the PR #2454 into master. Thank you @webdevilopers, @krewetka, @loconox for co-operation on this feature.

**Comment by webdevilopers** [comment]
Great work @pulzarraider . And it looks like the issue I mentioned regarding the naming of the select2 `input` in this comment
https://github.com/sonata-project/SonataAdminBundle/issues/2360#issuecomment-62383815
was also fixed.

There is an additional input generated for the `select2` input which has an uncommon naming `s54e1bb6fbe09f[supplierBranch]_autocomplete_input` - don't know if this is valid HTML or has to be anyway.
```twig
<input id="s54e1bb6fbe09f_supplierBranch_autocomplete_input" class="select2-offscreen" type="text" required="required" value="" name="s54e1bb6fbe09f[supplierBranch]_autocomplete_input" tabindex="-1" title="">
<div id="s54e1bb6fbe09f_supplierBranch_hidden_inputs_wrap">
<input type="hidden" value="" name="s54e1bb6fbe09f[supplierBranch]">
</div>
```

Is it generated by `select2` or the autocompleter? Maybe the brackets should be removed? No serious issue anyway I guess.

**Comment by rccc** [comment]
Hello,

Is this 2.4 version stable enough to update composer without break dependencies ?

**Comment by pulzarraider** [comment]
@webdevilopers the problem with input name was fixed in PR #2900

@rccc `2.4` was not released yet, but you can try it using `dev-master`. You should read the [UPGRADE-2.4.md](https://github.com/sonata-project/SonataAdminBundle/blob/master/UPGRADE-2.4.md).
